File: artifacts/libzahl-1.0/translate_new/src/zlsh.rs
use crate::*;
use std::ptr;

pub fn zlsh(a: &mut z_t, b: &z_t, bits: usize) {
    let mut chars: usize;
    let mut cbits: usize;
    let mut carry = [0, 0];
    
    if zzero(b) {
        a[0].sign = 0;
        return;
    }
    
    if bits == 0 {
        if a as *const _ != b as *const _ {
            zset(a, b);
        }
        return;
    }
    
    chars = bits >> 5;
    let bits = bits & (32 - 1);
    cbits = 32 - bits;
    
    if a[0].alloced < (b[0].used + chars) {
        libzahl_realloc(a, b[0].used + chars);
    }
    
    // Handle the chars array manipulation
    {
        let a_ptr = a as *const _;
        let b_ptr = b as *const _;
        let same_ptr = a_ptr == b_ptr;
        
        let a_chars = &mut a[0].chars;
        let b_chars = &b[0].chars;
        
        if let (Some(a_chars), Some(b_chars)) = (a_chars, b_chars) {
            // Ensure we have enough capacity
            if a_chars.len() < b[0].used + chars {
                libzahl_realloc(a, b[0].used + chars);
                // After realloc, we need to get a_chars again
                if let Some(a_chars) = &mut a[0].chars {
                    if same_ptr {
                        // Use memmove for overlapping regions
                        unsafe {
                            ptr::copy(
                                b_chars.as_ptr(),
                                a_chars.as_mut_ptr().add(chars),
                                b[0].used,
                            );
                        }
                    } else {
                        // Use memcpy for non-overlapping regions
                        unsafe {
                            ptr::copy_nonoverlapping(
                                b_chars.as_ptr(),
                                a_chars.as_mut_ptr().add(chars),
                                b[0].used,
                            );
                        }
                    }
                    
                    // Zero out the first chars elements
                    for i in 0..chars {
                        a_chars[i] = 0;
                    }
                    
                    a[0].used = b[0].used + chars;
                    
                    if bits != 0 {
                        for i in chars..a[0].used {
                            carry[(!i) & 1] = a_chars[i] >> cbits;
                            a_chars[i] <<= bits;
                            a_chars[i] |= carry[i & 1];
                        }
                        
                        let last_i = a[0].used - 1;
                        if carry[last_i & 1] != 0 {
                            if a[0].alloced < a[0].used + 1 {
                                libzahl_realloc(a, a[0].used + 1);
                                // After realloc, we need to get a_chars again
                                if let Some(a_chars) = &mut a[0].chars {
                                    // Create a new array with the additional element
                                    let mut new_chars = vec![0; a[0].used + 1];
                                    new_chars[..a[0].used].copy_from_slice(&a_chars[..a[0].used]);
                                    new_chars[a[0].used] = carry[last_i & 1];
                                    a[0].chars = Some(new_chars.into_boxed_slice());
                                    a[0].used += 1;
                                }
                            } else {
                                // Create a new array with the additional element
                                let mut new_chars = vec![0; a[0].used + 1];
                                new_chars[..a[0].used].copy_from_slice(&a_chars[..a[0].used]);
                                new_chars[a[0].used] = carry[last_i & 1];
                                a[0].chars = Some(new_chars.into_boxed_slice());
                                a[0].used += 1;
                            }
                        }
                    }
                }
            }
        }
    }
    
    a[0].sign = zsignum(b);
}

File: artifacts/libzahl-1.0/decompose/src/zlsh.c
void zlsh(z_t a, z_t b, size_t bits)
{
  size_t i;
  size_t chars;
  size_t cbits;
  zahl_char_t carry[] = {0, 0};
  if (zzero(b))
  {
    a->sign = 0;
    return;
  }
  if (!bits)
  {
    do
    {
      if (a != b)
      {
        zset(a, b);
      }
    }
    while (0);
    return;
  }
  chars = bits >> 5;
  bits = bits & (32 - 1);
  cbits = 32 - bits;
  do
  {
    if (a->alloced < (b->used + chars))
    {
      libzahl_realloc(a, b->used + chars);
    }
  }
  while (0);
  if (a == b)
  {
    memmove(a->chars + chars, b->chars, b->used * (sizeof(zahl_char_t)));
  }
  else
    memcpy(a->chars + chars, b->chars, b->used * (sizeof(zahl_char_t)));
  memset(a->chars, 0, chars * (sizeof(zahl_char_t)));
  a->used = b->used + chars;
  if (bits)
  {
    for (i = chars; i < a->used; i += 1)
    {
      carry[(~i) & 1] = a->chars[i] >> cbits;
      a->chars[i] <<= bits;
      a->chars[i] |= carry[i & 1];
    }

    if (carry[i & 1])
    {
      do
      {
        if (a->alloced < (a->used + 1))
        {
          libzahl_realloc(a, a->used + 1);
        }
      }
      while (0);
      a->chars[i] = carry[i & 1];
      a->used += 1;
    }
  }
  a->sign = zsignum(b);
}


/*** DEPENDENCIES:
void libzahl_realloc(z_t a, size_t need)
{
  size_t i;
  size_t x;
  zahl_char_t *new;
  unsigned int new_idx = 0;
  if ((need & ((~need) + 1)) != need)
  {
    need |= need >> 1;
    need |= need >> 2;
    need |= need >> 4;
    for (i = sizeof(need), x = 8; i; i >>= 1, x <<= 1)
      need |= need >> x;

    need += 1;
  }
  for (i = 0, x = need; x; x >>= 1)
    i += 1;

  if (libzahl_pool_n[i])
  {
    libzahl_pool_n[i] -= 1;
    new_idx = libzahl_pool[i][libzahl_pool_n[i]];
    memcpy(new, a->chars, a->alloced * (sizeof(zahl_char_t)));
    zfree(a);
    a->chars = &new[new_idx];
  }
  else
  {
    a->chars = realloc(a->chars, need * (sizeof(zahl_char_t)));
    if (!a->chars)
    {
      if (!errno)
      {
        errno = ENOMEM;
      }
      libzahl_error = errno, longjmp(libzahl_jmp_buf, 1);
    }
  }
  a->alloced = need;
}


----------------------------
typedef uint32_t zahl_char_t
----------------------------
void zset(z_t a, z_t b)
{
  if (zzero(b))
  {
    a->sign = 0;
  }
  else
  {
    do
    {
      if (a->alloced < b->used)
      {
        libzahl_realloc(a, b->used);
      }
    }
    while (0);
    a->sign = b->sign;
    a->used = b->used;
    memcpy(a->chars, b->chars, b->used * (sizeof(zahl_char_t)));
  }
}


----------------------------
inline static int zzero(z_t a)
{
  return !a->sign;
}


----------------------------
typedef struct 
{
  int sign;
  size_t used;
  size_t alloced;
  zahl_char_t *chars;
} z_t[1]
----------------------------
inline static int zsignum(z_t a)
{
  return a->sign;
}


----------------------------
***/


